woocommerce/woocommerce-ios
# Add custom view with refresh control in SwiftUI for Inbox

Part of #5954 


### Description


Since there is no native refresh control for `ScrollView` in SwiftUI (`refreshable` UI is only available for `List` and iOS 15+), I looked and tried a few "workaround options" online and chose [this simpler solution](https://developer.apple.com/forums/thread/654432?answerId=695019022#695019022). [This blog post](https://swiftuirecipes.com/blog/pull-to-refresh-with-swiftui-scrollview) is another close one, but a bit more complex. The UX is not the same as the native UIKit refresh control (you can compare with Reviews/Products pages), feel free to suggest other solutions. @pmusolino shared [this SwiftUI-Lab example](https://gist.github.com/swiftui-lab/3de557a513fbdb2d8fced41e40347e01) in p1644831260018489/1644830464.328879-slack-CGPNUU63E, but it seems to have some issues based on the latest comments.

New SwiftUI reusable views:
- `InfiniteScrollIndicator`: directly moved from `InfiniteScrollList`, it used to be private inside `InfiniteScrollList`
- `RefreshableScrollView`: adapted from [this Apple dev forum answer](https://developer.apple.com/forums/thread/654432?answerId=695019022#695019022) as mentioned earlier
- `RefreshableInfiniteScrollList`: similar to `InfiniteScrollList`, but with refresh control support using `RefreshableScrollView` instead of `ScrollView`

Lastly, I updated `Inbox` view to use `RefreshableInfiniteScrollList` instead of `InfiniteScrollList`.

Any suggestions to improve the UX are super welcome!

### Testing instructions


After implementing pull-to-refresh, I noticed that there are multiple notes with the same creation date (`dateCreated`) and the order is nondeterministic. I created a separate issue https://github.com/woocommerce/woocommerce-ios/issues/6271 to have a secondary sort descriptor on descending ID. Due to this issue, you might notice that some notes changed order after pulling to refresh.

Prerequisite: the site shows inbox notes in WC Admin (wp-admin > WooCommerce home page)
* Launch the app
* Go to the Menu tab
* Tap on "Inbox"
* After the list is loaded, pull down to refresh (right now the offset threshold is 50px) --> a spinner is shown to refresh the inbox notes. After the refresh is complete, the spinner should disappear

### Screenshots



https://user-images.githubusercontent.com/1945542/155071899-3b2a71ca-1b03-47fe-843a-f354f8473a78.mov



---
